NiL.JS/BaseLibrary/Console.cs

```diff
@@ -161,15 +187,235 @@ public JSValue log(Arguments args)
+
+            // top line
+            s.Append("+-");
+            for (int i = 0; i < columns; i++)
+            {
+                if (i > 0)
+                    s.Append("-+-");
+                int colWidth = cols[colsN[i]];
+                s.Append(new string('-', colWidth));
+            }
+            s.Append("-+\n");
+
+            // header
+            s.Append("| ");
+            for (int i = 0; i < columns; i++)
+            {
+                if (i > 0)
+                    s.Append(" | ");
+                int colWidth = cols[colsN[i]];
+                s.Append(colsN[i].PadRight(colWidth));
+            }
+            s.Append(" |\n");
+
+            // middle line
+            s.Append("+-");
+            for (int i = 0; i < columns; i++)
+            {
+                if (i > 0)
+                    s.Append("-+-");
+                s.Append(new string('-', cols[colsN[i]]));
+            }
+            s.Append("-+\n");
+
+            // body
+            for (int r = 0; r < rows.Count; r++)
+            {
+                var row = rows[r];
+                bool hasNextLine = true;
+                int offset = 0;
+                while (hasNextLine)
+                {
+                    hasNextLine = false;
+
+                    s.Append("| ");
+                    for (int i = 0; i < columns; i++)
+                    {
+                        if (i > 0)
+                            s.Append(" | ");
+                        string colName = colsN[i];
+                        int colWidth = cols[colName];
+                        string line = "";
+
+                        if (row.ContainsKey(colName))
+                        {
+                            var lines = row[colName];
+                            if (offset < lines.Length)
+                                line = lines[offset];
+                            if (lines.Length > offset + 1)
+                                hasNextLine = true;
+                        }
+
+                        s.Append(line.PadRight(colWidth));
+                    }
+                    s.Append(" |\n");
+
+                    offset++;
+                }
+            }
+
+            // bottom line
+            s.Append("+-");
+            for (int i = 0; i < colsN.Count; i++)
+            {
+                if (i > 0)
+                    s.Append("-+-");
+                s.Append(new string('-', cols[colsN[i]]));
+            }
+            s.Append("-+");
+
+
+            LogMessage(LogLevel.Log, s.ToString());
+
+            return JSValue.undefined;
+        }
+
+        public JSValue trace(Arguments args)
+        {
+            Context c = Context.CurrentContext;
+
+            StringBuilder s = new StringBuilder();
+
+            int i = 0;
+            while (c != null)
+            {
+                if (c._parent == null) // GlobalContext
+                    break;
+                if (i++ > Context.MaxConcurentContexts) // prevent infinite loops
```

Comment:
This condition is incorrect. `Context.MaxConcurentContexts` does not limit the depth of callstack. This is the limitation of the number of threads.